application/forms/EntryForm.php

```diff
@@ -311,114 +317,233 @@ protected function assemble()
     public function addEntry(int $scheduleId): void
     {
         $data = $this->formValuesToDb();
+        $membershipHash = $this->getMembershipHashFromRecipients($scheduleId);
         $recipients = array_map(function ($recipient) {
             return explode(':', $recipient, 2);
         }, explode(',', $this->getValue('recipient')));
 
         $db = Database::get();
         $db->beginTransaction();
 
-        $db->insert('timeperiod', ['owned_by_schedule_id' => $scheduleId]);
-        $timeperiodId = $db->lastInsertId();
+        $members = ScheduleMember::on(Database::get())
+            ->with('timeperiod')
+            ->filter(
+                Filter::all(
+                    Filter::equal('timeperiod.owned_by_schedule_id', $scheduleId),
+                    Filter::equal('schedule_member.membership_hash', $membershipHash)
+                )
+            );
+
+        if ($members->count() > 0) {
+            $timeperiodId = $members->first()->timeperiod_id;
+        } else {
+            $db->insert('timeperiod', ['owned_by_schedule_id' => $scheduleId]);
+            $timeperiodId = $db->lastInsertId();
+        }
 
         $db->insert('timeperiod_entry', $data + ['timeperiod_id' => $timeperiodId]);
 
-        foreach ($recipients as list($type, $id)) {
-            if ($type === 'contact') {
-                $db->insert('schedule_member', [
-                    'schedule_id' => $scheduleId,
-                    'timeperiod_id' => $timeperiodId,
-                    'contact_id' => $id
-                ]);
-            } elseif ($type === 'group') {
-                $db->insert('schedule_member', [
-                    'schedule_id' => $scheduleId,
-                    'timeperiod_id' => $timeperiodId,
-                    'contactgroup_id' => $id
-                ]);
+        if ($members->count() === 0) {
+            $binaryMemberHash = (new Binary([]))->toDb($membershipHash, null, null);
+            foreach ($recipients as list($type, $id)) {
+                if ($type === 'contact') {
+                    $db->insert('schedule_member', [
+                        'schedule_id'     => $scheduleId,
+                        'timeperiod_id'   => $timeperiodId,
+                        'contact_id'      => $id,
+                        'membership_hash' => $binaryMemberHash
+                    ]);
+                } elseif ($type === 'group') {
+                    $db->insert('schedule_member', [
+                        'schedule_id'     => $scheduleId,
+                        'timeperiod_id'   => $timeperiodId,
+                        'contactgroup_id' => $id,
+                        'membership_hash' => $binaryMemberHash
+                    ]);
+                }
             }
         }
 
         $db->commitTransaction();
     }
 
+    /**
+     * Get membership hash for the recipients in timeperiod entry for the given schedule
```

Comment:
Please be more specific here and less verbose. You don't need to mention timeperiod entries here, that's obvious from the class already. Then, `get` imposes that the value may be be cached or so. At least, I don't see it fit with the implementation. It's not a getter. So maybe rename the method to `calculate...` or `generate...`, or even `hash...`.